### PR TITLE
Code blocks should be editable after saving the post and reloading the editor

### PR DESCRIPTION
**Pull request by davilera**
## Description
See #13218.

## How has this been tested?
Created the block described in #13218.

## Types of changes
The PR solves the issue by making sure the content of a code block is pulled as HTML during the parsing process and escaping/unescaping tag delimiters (i.e. `<` and `>`) properly.

~Unfortunately, this seems to be (yet another) breaking change, as it requires `<` and `>` to be escaped before saving.~ **Edit:** 1ac7cfafd547231fe9920938f59c275e8bcbb97f prevents this PR from being a breaking change.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/block-library/src/code/index.js [hunk 1]
```diff
@@ -23,6 +23,13 @@ export const settings = {
 	supports: {
 		html: false,
 	},
+	attributes: {
```
Comment (aduth) [inline packages/block-library/src/code/index.js]: As part of the Block Registration RFC (#13693), attributes definitions for core blocks have been moved to `block.json` within the block directory (see #14770). Updates should be made there instead.

**Comment by aduth** [comment]
> Unfortunately, this seems to be (yet another) breaking change, as it requires `<` and `>` to be escaped before saving.

Can you elaborate on what specifically is breaking? Will existing content be shown as ["Invalid"](https://developer.wordpress.org/block-editor/developers/block-api/block-edit-save/#validation)?

File: packages/block-library/src/code/index.js (hunk 1 above)
Comment (davilera) [reply packages/block-library/src/code/index.js]: Ok, thanks! I'll fix this tomorrow first thing in the morning.

**Comment by davilera** [comment]
> Can you elaborate on what specifically is breaking? Will existing content be shown as "Invalid"?

I think it would, yes. However, there's an option to make a block compatible with previous versions, right? If so, I can explore this...

**Comment by aduth** [comment]
It'd be worth testing to see if it is an issue. The block validation logic explicitly includes some leniency where differences are purely on differences in encoding ([source](https://github.com/WordPress/gutenberg/blob/f1bbc004734ea157dd72ed5bf1f808c10efc0e84/packages/blocks/src/api/validation.js#L222-L243)), so it might not be a problem.

If it does turn out to be breaking, we can try to use the `deprecated` feature for transitioning ([reference](https://developer.wordpress.org/block-editor/developers/block-api/block-deprecation/).

**Comment by davilera** [comment]
OK so... I think I got this (without introducing breaking changes!). The problem was, Gutenberg (or the browser/JavaScript itself, I don't know) escapes "less than" characters (`<`) when the block is saved in the database. That is, something like #13218:

```html
<a href="http://wordpress.org/">WordPress</a>  &lt;strong&gt;
```

becomes this (after my original PR #13996 is applied):

```html
&lt;a href="http://wordpress.org/">WordPress&lt;/a>  &amp;lt;strong&amp;gt;
```

The problem is, when loading that block from the post content, characters like `&lt;` aren't properly parsed and the problems described in #13218 arise. I tweaked the block so that:

1. Content is retrieved in HTML, so that the block can work with HTML entities.
2. The `unescape` function we use to set the value in the `PlainText` component also unescapes `&lt;` and `&gt;` characters, which might be there because of the save process.

Apparently, this seems to solve issue #13218 and all tests pass. I would recommend a few more testing, though.

**Comment by davilera** [comment]
Were you able to review this, @aduth? Is there anything else I can do to help?

**Comment by aduth** [comment]
> 2\. The `unescape` function we use to set the value in the `PlainText` component also unescapes `&lt;` and `&gt;` characters, which might be there because of the save process.

Yep, wondering about why we wouldn't need to have a corresponding escaper in the setter, it occurs to me that this is already effectively escaped when the block is serialized. So I think the change is sensible in restoring a balance. It's also better for it to occur during the save than in `escape` since if we were to escape the value as maintained in the block attributes, it would wrongly be displayed as escaped in the `PlainText` textarea.

I'd wondered about whether we really need to unescape here, vs. setting the `innerHTML` of the `textarea` directly, since this is a [common technique](https://stackoverflow.com/a/7394787) for "safely" decoding entities. That said, I'm inclined to avoid `dangerouslySetInnerHTML` as much as possible for reasons which should be self-evident by its name 😄 My only concern then is if there are other encoded entities which we need to be considering.

This looks good to me. I'll want to give it another pass in the morning with a fresher mind.

**Comment by aduth** [comment]
> Yep, wondering about why we wouldn't need to have a corresponding escaper in the setter, it occurs to me that this is already effectively escaped when the block is serialized. So I think the change is sensible in restoring a balance.

In further confirming this, I expect it should be effectively the exact opposite of:

https://github.com/WordPress/gutenberg/blob/13e58516cc34dfd3fabc569ef2177c332091964e/packages/escape-html/src/index.js#L95-L97

There are two observations:

- We need ampersand unescaping. _This already existed prior to the pull request_.
- We technically _don't need_ "greater than" unescaping. I expect it should cause no harm to keep it, however, since this unescaping is only meant to account for escaping we only expect to occur from the above serialization logic anyways.

I'd also wondered about server-side escaping we might need to anticipate, and specifically in cases of [`unfiltered_html`](https://wordpress.org/support/article/roles-and-capabilities/#unfiltered_html) for non-privileged users. In investigating how this is applied, it seems most relevant for tag-stripping and not for character escaping ([reference](https://developer.wordpress.org/reference/functions/wp_filter_post_kses/)).

**Comment by aduth** [comment]
@davilera I have some small concern about trying to keep this in sync with the specific implementation of the serializer, and I'm curious your thoughts on an alternative I have in mind:

The idea would be to change the `source` to `'html'` (like as proposed) and use `dangerouslySetInnerHTML` in the `save` implementation to bypass the serializer escaping. To avoid having characters shown as escaped in the textarea, we would need to either use `dangerouslySetInnerHTML` in the `edit` implementation (which is a bit less "safe" in my mind than in `save`), or another way to decode the entities before assigning the `value`. There is a [`@wordpress/html-entities`](https://github.com/WordPress/gutenberg/tree/master/packages/html-entities) package for this purpose ([its implementation](https://github.com/WordPress/gutenberg/blob/master/packages/html-entities/src/index.js) is admittedly not far from as the alternative). The attribute value would need to be maintained in its escaped form, so as to align with the new `save` implementation. We could use the textarea's `innerHTML` for this, which as far as I can tell will take care of this.

```js
var e = document.createElement( 'textarea' ); e.textContent = '&'; console.log( e.innerHTML );
// '&amp;'
```

It seems this alternative might pose some more risk in trying to assure safety of value escaping, so I may be comfortable as well with the current proposal, even if it needs to account for factors outside its own consideration (the serializer behavior).

**Comment by davilera** [comment]
> To avoid having characters shown as escaped in the textarea, we would need to either use `dangerouslySetInnerHTML` in the edit implementation (which is a bit less "safe" in my mind than in save), or another way to decode the entities before assigning the value.

To be honest, I don't have a strong opinion on any solution, as none is perfect and each poses its own problems. That said, I feel like using `dangerouslySetInnerHTML` is slightly worse (I'm not sure we'd be able to control what it does).

Regarding the current proposal, I also don't like using an asymmetric `escape` and `unescape` functions in [utils.js](https://github.com/WordPress/gutenberg/blob/1ac7cfafd547231fe9920938f59c275e8bcbb97f/packages/block-library/src/code/utils.js#L12) because, as @aduth said, “it needs to account for factors outside its own consideration (the serializer behavior)”. If we make it symmetric, the block should work (I think), but we'd be introducing a breaking change (as we'd be changing what's saved in the DB).

What do you think, @aduth? Should we introduce a breaking change? I think the risk the current proposal poses is minimal: in principle, our block would only receive a `&lt;` character in `content` if somebody else (e.g. the serializer) inserted it. If it was the user who wrote the `&lt;` entity, this would be saved as `&amp;lt;` (because of our `[escape](https://github.com/WordPress/gutenberg/blob/1ac7cfafd547231fe9920938f59c275e8bcbb97f/packages/block-library/src/code/utils.js#L12)` method), which means things should work as expected. But I'd rather have a self-contained solution... Complicated decisions here!

**Comment by davilera** [comment]
If I'm not mistaken, the issue this PR addressed has been fixed by @ellatrix in #17994, so I'm closing it.

**Pull request closed without merging**